### PR TITLE
MINIFICPP-994 Move out OpenCV build from cmake configuration

### PR DESCRIPTION
**Pull request by lordgamez**
Jira ticket: https://issues.apache.org/jira/browse/MINIFICPP-994

----------------------------------------------------------------------------------------------
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Review by szaszm — COMMENTED** [review]
Thank you for cleaning this up. 2 minor comments below.

File: cmake/BundledOpenCV.cmake [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    else()
+        set(OPENCV_INCLUDE_DIR "${OPENCV_BYPRODUCT_DIR}/include/opencv4" CACHE STRING "" FORCE)
+    endif()
+    set(OPENCV_LIBRARIES "${OPENCV_LIBS}" CACHE STRING "" FORCE)
+
+    # Set exported variables for FindPackage.cmake
+    set(PASSTHROUGH_VARIABLES ${PASSTHROUGH_VARIABLES} "-DEXPORTED_OPENCV_INCLUDE_DIR=${OPENCV_INCLUDE_DIR}" CACHE STRING "" FORCE)
+    string(REPLACE ";" "%" OPENCV_LIBRARIES_EXPORT "${OPENCV_LIBRARIES}")
+    set(PASSTHROUGH_VARIABLES ${PASSTHROUGH_VARIABLES} "-DEXPORTED_OPENCV_LIBRARIES=${OPENCV_LIBRARIES_EXPORT}" CACHE STRING "" FORCE)
+    message(${PASSTHROUGH_VARIABLES})
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply cmake/BundledOpenCV.cmake]: Thanks, I left i there by accident when debugging, removed in 18fb02160b50f0c461b512e799d835d55b81a6e5

File: cmake/BundledOpenCV.cmake [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        "${OPENCV_BYPRODUCT_DIR}/${CMAKE_INSTALL_LIBDIR}/${PREFIX}opencv_photo${SUFFIX}"
+        "${OPENCV_BYPRODUCT_DIR}/${CMAKE_INSTALL_LIBDIR}/${PREFIX}opencv_videoio${SUFFIX}"
+        "${OPENCV_BYPRODUCT_DIR}/${CMAKE_INSTALL_LIBDIR}/${PREFIX}opencv_video${SUFFIX}"
+        "${OPENCV_BYPRODUCT_DIR}/${CMAKE_INSTALL_LIBDIR}/${PREFIX}opencv_stitching${SUFFIX}"
+        "${OPENCV_BYPRODUCT_DIR}/${CMAKE_INSTALL_LIBDIR}/${PREFIX}opencv_features2d${SUFFIX}"
+        "${OPENCV_BYPRODUCT_DIR}/${CMAKE_INSTALL_LIBDIR}/${THIRDPARTY_DIR}${PREFIX}libjpeg-turbo${THIRDPARTY_SUFFIX}"
+        "${OPENCV_BYPRODUCT_DIR}/${CMAKE_INSTALL_LIBDIR}/${THIRDPARTY_DIR}${PREFIX}libpng${THIRDPARTY_SUFFIX}")
+
+    # Set build options
+    set(OPENCV_CMAKE_ARGS ${PASSTHROUGH_CMAKE_ARGS}
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply cmake/BundledOpenCV.cmake:60]: Added in 18fb02160b50f0c461b512e799d835d55b81a6e5

**Review by szaszm — APPROVED** [review]

**Review by fgerlits — COMMENTED** [review]
LGTM, with some very minor comments

File: CMakeLists.txt [hunk 3]
```diff
@@ -503,6 +517,8 @@ endif()
 ## OpenCV Extensions
 option(ENABLE_OPENCV "Disables the OpenCV extensions." OFF)
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply CMakeLists.txt]: Updated in db9827c54077e213ac5c4f1c4ce871d155c0c40d

File: CMakeLists.txt [hunk 4]
```diff
@@ -503,6 +517,8 @@ endif()
...
 if (ENABLE_OPENCV)
+	include(BundledOpenCV)
+	use_bundled_opencv(${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR})
 	createExtension(OPENCV-EXTENSIONS "OPENCV EXTENSIONS" "This enabled OpenCV support" "extensions/opencv" "extensions/opencv/tests")
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply CMakeLists.txt]: Updated in db9827c54077e213ac5c4f1c4ce871d155c0c40d

**Review by martinzink — APPROVED** [review]

**Review by szaszm — CHANGES_REQUESTED** [review]
This PR causes a build issue:
```
+ nice -n 19 ninja -v -j64
ninja: error: 'thirdparty/opencv-install/lib64/libopencv_flann.a', needed by 'main/minifi', missing and no known rule to make it
```
May or may not happen with GNU Make, I only tried building with Ninja on Linux (with both GCC and Clang) after my merge attempt. Will postpone merging until this is fixed.

My cmake flags:
```
cmake -G Ninja -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DFORCE_COLORED_OUTPUT=ON -DAWS_ENABLE_UNITY_BUILD=OFF -DASAN_BUILD=OFF -DCMAKE_BUILD_TYPE=Debug -DFAIL_ON_WARNINGS=ON -DENABLE_PYTHON=ON -DENABLE_OPS=ON -DENABLE_JNI=OFF -DENABLE_OPC=ON -DENABLE_COAP=ON -DENABLE_GPS=ON -DENABLE_MQTT=ON -DENABLE_LIBRDKAFKA=ON -DENABLE_SENSORS=ON -DENABLE_USB_CAMERA=ON -DENABLE_AWS=OFF -DENABLE_SFTP=ON -DENABLE_OPENWSMAN=ON -DENABLE_BUSTACHE=ON -DENABLE_OPENCV=ON -DENABLE_TENSORFLOW=OFF -DENABLE_SQL=OFF -DENABLE_PCAP=OFF -DENABLE_NANOFI=ON -DENABLE_SYSTEMD=ON ..
```
Same for clang, except substitute gcc with clang and g++ with clang++.

**Comment by lordgamez** [comment]
> This PR causes a build issue:
> 
> ```
> + nice -n 19 ninja -v -j64
> ninja: error: 'thirdparty/opencv-install/lib64/libopencv_flann.a', needed by 'main/minifi', missing and no known rule to make it
> ```
> 
> May or may not happen with GNU Make, I only tried building with Ninja on Linux (with both GCC and Clang) after my merge attempt. Will postpone merging until this is fixed.
> 
> My cmake flags:
> 
> ```
> cmake -G Ninja -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DFORCE_COLORED_OUTPUT=ON -DAWS_ENABLE_UNITY_BUILD=OFF -DASAN_BUILD=OFF -DCMAKE_BUILD_TYPE=Debug -DFAIL_ON_WARNINGS=ON -DENABLE_PYTHON=ON -DENABLE_OPS=ON -DENABLE_JNI=OFF -DENABLE_OPC=ON -DENABLE_COAP=ON -DENABLE_GPS=ON -DENABLE_MQTT=ON -DENABLE_LIBRDKAFKA=ON -DENABLE_SENSORS=ON -DENABLE_USB_CAMERA=ON -DENABLE_AWS=OFF -DENABLE_SFTP=ON -DENABLE_OPENWSMAN=ON -DENABLE_BUSTACHE=ON -DENABLE_OPENCV=ON -DENABLE_TENSORFLOW=OFF -DENABLE_SQL=OFF -DENABLE_PCAP=OFF -DENABLE_NANOFI=ON -DENABLE_SYSTEMD=ON ..
> ```
> 
> Same for clang, except substitute gcc with clang and g++ with clang++.
> 
> These are the exact scripts I use to test builds: https://github.com/szaszm/verify-nifi-release/tree/master/bin

Thanks for all the info! There was an error in the cmake file, the wrong variable was passed to the build byproduct parameter. It worked fine with Make, but ninja fails without defining the byproducts beforehand. I fixed the issue in f0ab4c23ec482140065bd51a13b71eaa98900f71

**Review by szaszm — APPROVED** [review]

**Pull request closed without merging**